src/lib/storage/dictionary_segment.cpp

```diff
@@ -18,7 +18,11 @@ DictionarySegment<T>::DictionarySegment(const std::shared_ptr<const pmr_vector<T
       _dictionary{dictionary},
       _attribute_vector{attribute_vector},
       _null_value_id{null_value_id},
-      _decompressor{_attribute_vector->create_base_decompressor()} {}
+      _decompressor{_attribute_vector->create_base_decompressor()} {
+  // We need the last possible value ID to symbolize NULL in the attribute vector and "value not found" in
```

Comment:
Regarding the representation of NULL: Let's think of a dictionary with 10 values from 1 to 10 again. The data type of ValueID would be `uint8_t`. I understand the last possible ValueID to symbolize NULL to be `std::numeric_limits<uint8_t>::max()` or `256`.

However, the comment below says that NULL is represented with the ValueID of `unique_values_count()` which should be `10`.

Reply:
How about this:

> NULL is represented by `_dictionary.size()`. `INVALID_VALUE_ID`, which is the highest possible number in ValueID::base_type (`2^32 -1`) is needed to represent "value not found" in calls to `lower_bound/upper_bound`. For a DictionarySegment of the max size `Chunk::MAX_SIZE`, those two values overlap.

```
Assert(_dictionary->size() == _null_value_id), "Invalid NULL value id");
Assert(_dictionary->size() < std::numeric_limits<ValueID::base_type>::max(), "Input segment too big");
```